### PR TITLE
[5.8] Nested keys for Higher Order Proxy

### PR DESCRIPTION
**Pull request by nmokkenstorm**
Hey all,

I ran into a minor issue earlier this morning. I wanted to consume an JSON API response in the following format:

`[
[
'values' => ['value' => 100],
'time_range' => ['date' => '04-03-2019']
],
[
'values' => ['value' => 200], 
'time_range' => ['date' => '04-02-2019']
]
]`

I assumed I could keyBy using dot notation in Higher Order Proxy, because you can using the normal collection. I wrote an extra test to confirm this, but the relevant code is in line 1781 of Collection. Collection uses data_get, and HigherOrderCollectionProxy uses direct array access.

Obviously I could just use $collection->keyBy('time_range.date') here, but I deeply enjoy the brevity HAP provides and I'd love to see it used more.

Long story short, this PR should make HAP and regular Collections more consistent, and I feel that consuming deeply nested arrays is a regular enough use-case to warrant this, it shouldn't break any existing behaviour, so I targeted 5.8, let me know if that's wrong.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by nmokkenstorm** [comment]
Hello Taylor,

I understand the decision, the proposed change was more about achieving more consistency in the behaviour in higher order messaging and Collection, but not having it is neither a bug nor a show-stopping annoyance for now. Thanks for having a look at it anyways!
